src/charm.py

```diff
@@ -191,185 +190,101 @@ def unit_peer_data(self) -> Dict:
+            return value
 
-        # It's not the first secret for the scope, we can reuse the existing one
-        # that was fetched in the previous call
-        if secret:
-            secret_cache = self.secrets[scope][SECRET_CACHE_LABEL]
-
-            if secret_cache.get(key) == value:
-                logging.debug(f"Key {scope}:{key} has this value defined already")
-            else:
-                secret_cache[key] = value
-                try:
-                    secret.set_content(secret_cache)
-                except OSError as error:
-                    logging.error(
-                        f"Error in attempt to set {scope}:{key}. "
-                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
-                    )
-                    return
-                logging.debug(f"Secret {scope}:{key} was {key} set")
-
-        # We need to create a brand-new secret for this scope
-        else:
-            scope_obj = self._scope_obj(scope)
+        if JujuVersion.from_environ().has_secrets:
+            secret_key = self._translate_field_to_secret_key(key)
+            label = generate_secret_label(self, scope)
+            for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(1), reraise=True):
+                with attempt:
+                    secret = self.secrets.get(label)
 
-            secret = scope_obj.add_secret({key: value})
             if not secret:
-                raise RuntimeError(f"Couldn't set secret {scope}:{key}")
-
-            self.secrets[scope][SECRET_LABEL] = secret
-            self.secrets[scope][SECRET_CACHE_LABEL] = {key: value}
-            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
-            peer_data.update({SECRET_INTERNAL_LABEL: secret.id})
-
-        # TODO change upgrade to switch to secrets once minor version upgrades is done
-        if key in peer_data:
-            del peer_data[key]
+                return
 
-        return self.secrets[scope][SECRET_LABEL].id
+            value = secret.get_content().get(secret_key)
+            if value != SECRET_DELETED_LABEL:
+                return value
 
-    def set_secret(self, scope: str, key: str, value: Optional[str]) -> Optional[str]:
+    def set_secret(self, scope: Scopes, key: str, value: Optional[str]) -> Optional[str]:
         """Set secret from the secret storage."""
-        if scope not in [APP_SCOPE, UNIT_SCOPE]:
+        if scope not in get_args(Scopes):
             raise RuntimeError("Unknown secret scope.")
 
         if not value:
             return self.remove_secret(scope, key)
 
-        juju_version = JujuVersion.from_environ()
+        if JujuVersion.from_environ().has_secrets:
+            if self._peer_data(scope).get(key):
+                # Charm must have been upgraded since last run
+                # We move from databag to secrets
+                self._peer_data(scope).pop(key)
```

Comment:
```suggestion
            # Charm must have been upgraded since last run
            # We move from databag to secrets
            self._peer_data(scope).pop(key, None)
```